### ISSUE TITLE
Sequential load fraction bug w/ zonal UnitarySystem equipment

### ISSUE THREAD

**Issue by shorowit** [NREL/EnergyPlus#7264]
Issue overview
--------------
I'm testing a building in E+ 9.1 that has a zonal unitary system plus an ideal air system, each serving 50% of the heating load using either UniformLoad or SequentialLoad. However, the simulation results for the SequentialLoad simulations are incorrectly varying depending on Zone Equipment Sequence order. One of the SequentialLoad runs is perfectly matching the UniformLoad case while the other is not. (I've separately confirmed that the UniformLoad results are correct by comparing them to a building with the zonal unitary system meeting 100% of the heating load.)

Interestingly enough, I'm not seeing this issue with other zonal equipment (e.g., PTACs) or unitary systems on air loops. It's only showing up for zonal unitary systems.

Files:
[UniformLoad-good.idf.txt](https://github.com/NREL/EnergyPlus/files/3061448/UniformLoad-good.idf.txt)
[SequentialLoad-good.idf.txt](https://github.com/NREL/EnergyPlus/files/3061451/SequentialLoad-good.idf.txt)
[SequentialLoad-bad.idf.txt](https://github.com/NREL/EnergyPlus/files/3061452/SequentialLoad-bad.idf.txt)

IDF | Heating [GJ] | Heating Fraction
-- | -- | --
UniformLoad-good.idf | 17.91 | 0.50
SequentialLoad-good.idf | 17.91 | 0.50
SequentialLoad-bad.idf | 12.58 | 0.35

@nmerket @mjwitte 

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

**Comment by rraustad** [comment]
Does this have something to do with not allocating Cooling/Heating fraction to 0.5 for both equipment types? I guess regardless the loads should be met either way.

SequentialLoad-good.idf

    ZoneHVAC:EquipmentList,
      living Equipment List,                  !- Name
      SequentialLoad,                         !- Load Distribution Scheme
      AirLoopHVAC:UnitarySystem,              !- Zone Equipment Object Type 1
      res unit heater unitary system,         !- Zone Equipment Name 1
      2,                                      !- Zone Equipment Cooling Sequence 1
      2,                                      !- Zone Equipment Heating or No-Load Sequence 1
      1,                                      !- Zone Equipment Sequential Cooling Fraction 1
      1,                                      !- Zone Equipment Sequential Heating Fraction 1
      ZoneHVAC:IdealLoadsAirSystem,           !- Zone Equipment Object Type 2
      res ideal heat,                         !- Zone Equipment Name 2
      1,                                      !- Zone Equipment Cooling Sequence 2
      1,                                      !- Zone Equipment Heating or No-Load Sequence 2
      1,                                      !- Zone Equipment Sequential Cooling Fraction 2
      0.5;                                    !- Zone Equipment Sequential Heating Fraction 2

SequentialLoad-bad.idf

    ZoneHVAC:EquipmentList,
      living Equipment List,                  !- Name
      SequentialLoad,                         !- Load Distribution Scheme
      AirLoopHVAC:UnitarySystem,              !- Zone Equipment Object Type 1
      res unit heater unitary system,         !- Zone Equipment Name 1
      1,                                      !- Zone Equipment Cooling Sequence 1
      1,                                      !- Zone Equipment Heating or No-Load Sequence 1
      1,                                      !- Zone Equipment Sequential Cooling Fraction 1
      0.5,                                    !- Zone Equipment Sequential Heating Fraction 1
      ZoneHVAC:IdealLoadsAirSystem,           !- Zone Equipment Object Type 2
      res ideal heat,                         !- Zone Equipment Name 2
      2,                                      !- Zone Equipment Cooling Sequence 2
      2,                                      !- Zone Equipment Heating or No-Load Sequence 2
      1,                                      !- Zone Equipment Sequential Cooling Fraction 2
      1;                                      !- Zone Equipment Sequential Heating Fraction 2

**Comment by shorowit** [comment]
The unitary system only has a heating coil and the ideal air system can only provide heating (cooling capacity is set to 0), so I would hope not.

**Comment by rraustad** [comment]
I also see this. The cooling and heating loads are multiplied by the same ratio (loadRatio) yet these ratios are different for cooling and heating. I don't see how this would make a difference but it stands out as a possible place to look. I'll try using the same fractions to see what happens with the bad test file.

                // Determine whether we're heating or cooling and choose the appropriate fraction
                const Real64 &heatLoadRatio = thisZEqList.SequentialHeatingFraction(equipNum);
                const Real64 &coolLoadRatio = thisZEqList.SequentialCoolingFraction(equipNum);
                const Real64 loadRatio = (energy.TotalOutputRequired >= 0.0) ? heatLoadRatio : coolLoadRatio;

                // Energy loads
                energy.SequencedOutputRequired(priorityNum) = energy.TotalOutputRequired * loadRatio;
                energy.SequencedOutputRequiredToHeatingSP(priorityNum) = energy.OutputRequiredToHeatingSP * loadRatio;
                energy.SequencedOutputRequiredToCoolingSP(priorityNum) = energy.OutputRequiredToCoolingSP * loadRatio;
                energy.RemainingOutputRequired = energy.SequencedOutputRequired(priorityNum);
                energy.RemainingOutputReqToHeatSP = energy.SequencedOutputRequiredToHeatingSP(priorityNum);
                energy.RemainingOutputReqToCoolSP = energy.SequencedOutputRequiredToCoolingSP(priorityNum);

**Comment by rraustad** [comment]
For the bad test file I get 17.91 GJ for heating electricity in the End Use table. So this seems to work if fractions are consistent. The thing is I don't know if this is a workaround or if something else should have happened when the fractions weren't consistent. Need @mjwitte input.

    ZoneHVAC:EquipmentList,
      living Equipment List,                  !- Name
      SequentialLoad,                         !- Load Distribution Scheme
      AirLoopHVAC:UnitarySystem,              !- Zone Equipment Object Type 1
      res unit heater unitary system,         !- Zone Equipment Name 1
      1,                                      !- Zone Equipment Cooling Sequence 1
      1,                                      !- Zone Equipment Heating or No-Load Sequence 1
      0.5,                                      !- Zone Equipment Sequential Cooling Fraction 1
      0.5,                                    !- Zone Equipment Sequential Heating Fraction 1
      ZoneHVAC:IdealLoadsAirSystem,           !- Zone Equipment Object Type 2
      res ideal heat,                         !- Zone Equipment Name 2
      2,                                      !- Zone Equipment Cooling Sequence 2
      2,                                      !- Zone Equipment Heating or No-Load Sequence 2
      0.5,                                      !- Zone Equipment Sequential Cooling Fraction 2
      0.5;                                      !- Zone Equipment Sequential Heating Fraction 2

**Comment by rraustad** [comment]
I think the next step would be to see what portion of the zone load each of these systems is meeting. I assume the intention is for each to meet 1/2 of the load.

**Comment by shorowit** [comment]
Interesting sleuthing, @rraustad. Note that the Sequential Fractions are fractions of **remaining** load (i.e., after the previous equipment has served part of the load). So your scenario above actual means that the unitary system serves 50% of the heating load and the ideal air system meets 50% of the remaining load (so 25% of the heating load). I'm surprised that gives the correct result???

If I update your scenario with Sequential Fractions of 1 for the 2nd equipment, the results are once again incorrect.

And yes, the intention is for each to meet 1/2 the load.

**Comment by shorowit** [comment]
The results so far seem to confirm that the problem is not with the Cooling Fractions. Their values (correctly) have no impact on the results given that the building only has heating equipment.

**Comment by rraustad** [comment]
The IDD shows the fraction is applied (or implied) to the remaining load but the code is using a fraction of the total (OutputRequiredToHeatingSP) zone load so I am surprised either of these worked:

    energy.RemainingOutputReqToHeatSP = energy.OutputRequiredToHeatingSP * loadRatio

    N4 , \field Zone Equipment 1 Sequential Heating Fraction
         \note The fraction of the remaining heating load this equipment will attempt to serve
         \note if the load distribution scheme is SequentialLoad, otherwise ignored.

**Comment by rraustad** [comment]
Here's what the total delivered load (light blue and orange) and also what each system is doing for the good and bad sequential load cases (Mdot*DeltaH from supply node to zone node). Looks to me as if:
1) the IDD note is misleading and the sequential fraction is a fraction of the total zone load
2) the sequential fractions should add up to 1 

![image](https://user-images.githubusercontent.com/5778156/55899462-e06b6c00-5b92-11e9-9e51-c5ccb7a30c10.png)

**Comment by rraustad** [comment]
This is somewhat confusing (which load variable to use when) and I think the last piece of the puzzle.

PurcahsedAirManager always uses:

        // get current zone requirements
        QZnHeatSP = ZoneSysEnergyDemand(ActualZoneNum).RemainingOutputReqToHeatSP;
        QZnCoolSP = ZoneSysEnergyDemand(ActualZoneNum).RemainingOutputReqToCoolSP;

And UnitarySystem uses (enters the IF part with these test files and probably all other files as well since they all use EquipLists):

        if (this->m_ZoneSequenceCoolingNum > 0 && this->m_ZoneSequenceHeatingNum > 0) {
            QToCoolSetPt = DataZoneEnergyDemands::ZoneSysEnergyDemand(this->ControlZoneNum)
                                       .SequencedOutputRequiredToCoolingSP(this->m_ZoneSequenceCoolingNum);
            QToHeatSetPt = DataZoneEnergyDemands::ZoneSysEnergyDemand(this->ControlZoneNum)
                                       .SequencedOutputRequiredToHeatingSP(this->m_ZoneSequenceHeatingNum);
        } else {
            ZoneLoad = DataZoneEnergyDemands::ZoneSysEnergyDemand(this->ControlZoneNum).RemainingOutputRequired;
            QToCoolSetPt = DataZoneEnergyDemands::ZoneSysEnergyDemand(this->ControlZoneNum).OutputRequiredToCoolingSP;
            QToHeatSetPt = DataZoneEnergyDemands::ZoneSysEnergyDemand(this->ControlZoneNum).OutputRequiredToHeatingSP;
            MoistureLoad = DataZoneEnergyDemands::ZoneSysMoistureDemand(this->ControlZoneNum).OutputRequiredToDehumidifyingSP;
        }

**Comment by mjwitte** [comment]
> the IDD note is misleading and the sequential fraction is a fraction of the total zone load
the sequential fractions should add up to 1

1. No, the design intent is for the fraction to be applied to the *remaining* load. The fractions should not add up to 1.

2. `DistributeSystemOutputRequired` is only called for the first equipment, and only the first sequenced load is set to `TotalOutputRequired * loadRatio`

3. After that, `UpdateSystemOutputRequired` takes over. 
There all of the `remainingOutpuReq*` variables are updated to `loadRatio * energy.UnadjRemainingOutputReq*`

4. AirloopHVAC:UnitarySystem as zone equipment should be using **Remaining**OutputRequiredToCoolingSP, etc.

**Comment by rraustad** [comment]
OK, but the IdealLoadsAirSystem IS using RemainingOut;putRequired. In the bad test file this equipment is the second in the list. The UnitarySystem is 1st and met 1/2 the load (see the figure above, dark blue on top of yellow/green). So why didn't the IdealLoads system meet the other half (it met the whole load, the red line right on top of light blue line)?

Also, I don't like using loadRatio since that assumes the equipment will be in that mode of operation (heating or cooling). The UnitarySystem can flip the load (or operating mode) based on OA impact. It would need to accurately know the load to the opposite set point. If there is a heating load, and the UnitarySys flipped the load/operating mode, it needs to know the load to cooling SP, which would be multiplied by the heating fraction. If it were multiplied by the cooling fraction I think that would be OK.

**Comment by mjwitte** [comment]
Not sure about the defect file issue, I haven't looked at it yet.  I wonder, when unitary runs a zone equipment, is it setting `SysOutputProvided` ? No - it's not. `SysOutputProvided` and `LatOutputProvided` aren't even in the argument list in `SimZoneEquipment`:
```
                    } else if (SELECT_CASE_var == ZoneUnitarySys_Num) { // 'AirloopHVAC:UnitarySystem'
                        int AirLoopNum = 0;
                        bool HeatingActive = false;
                        bool CoolingActive = false;
                        int OAUnitNum = 0;
                        Real64 OAUCoilOutTemp = 0.0;
                        bool ZoneEquipFlag = true;
                        ZoneEquipList(CurZoneEqNum)
                            .compPointer[EquipPtr]
                            ->simulate(PrioritySimOrder(EquipTypeNum).EquipName,
                                       FirstHVACIteration,
                                       AirLoopNum,
                                       ZoneEquipList(CurZoneEqNum).EquipIndex(EquipPtr),
                                       HeatingActive,
                                       CoolingActive,
                                       OAUnitNum,
                                       OAUCoilOutTemp,
                                       ZoneEquipFlag);
```

The remaining load to the zone at the time the equipment is called is what matters. If the zone needs some heat, then meet loadRatio of that, no matter what kind of coils have to operate to meet it. @nmerket and I thought this through at the time and decided this logic was good. But we could be wrong.

[Above](https://github.com/NREL/EnergyPlus/issues/7264#issuecomment-481748113) you say Unitary is entering the IF? Meaning it's using the sequenced loads? Those should be ok in this case, but I'm surprised because the sequence numbers should only be set if it's connected to an airterminal. When it's zone equipment, I thought it wouldn't use the sequenced loads.

**Comment by nmerket** [comment]
Getting caught up here...

This code is a mess. It's not surprising to me that there is equipment that it doesn't work quite right for since each of the 20 or so different equipment types uses the load variables in slightly different ways. Without an intimate knowledge of how each works, it is nearly impossible to get this right. 

My intuition is that the "incorrect" application of the heating and cooling fractions isn't causing the problem here since in general it doesn't matter what cooling fraction you apply when you're in heating mode. (Yes, I can accept there are likely edge cases, but I don't think it would cause the issues we're seeing here.)

And yes, the fractions should not add up to one. It's confusing but permits it to work in some convenient ways.

**Comment by rraustad** [comment]
Good feedback. The biggest issue here is the UnitarySystem is not returning Sys/LatOutputProvided. I guess past user configurations never included other equipment in a zone. I see that air loop equipment don't return these either since the terminal unit takes care of that, so when I added ability to configure as zone equipment I missed that aspect. This is a good starting place.

Not sure I like adding an optional argument since I tried to eliminate those along the way. I guess I could pass a dummy and not use it for air loops. Can I pass a dummy argument that is initialized on the fly (i.e., without adding another line as
Real64 SysOutputProvided = 0.0 before the call to ::simulate)? Would this work?

    void UnitarySys::simulate(std::string const &Name, bool const FirstHVACIteration, int const &AirLoopNum,
                              int &CompIndex, bool &HeatActive, bool &CoolActive, int const ZoneOAUnitNum,
                              Real64 const OAUCoilOutTemp, bool const ZoneEquipment, Real64 SysOutputProvided(0.0), same for Lat)

**Comment by mjwitte** [comment]
How about a different function name when called as zone equipment? Or just a different signature?

**Comment by rraustad** [comment]
Passing back SysOutputProvided corrected the issue.

![image](https://user-images.githubusercontent.com/5778156/56687524-7b406c00-66a4-11e9-9392-47a8c5e89f3b.png)

**Comment by mjwitte** [comment]
Closed vis #7280
